### ISSUE TITLE
Multi-result tasks

### ISSUE THREAD

**Issue by k32** [k32/task_graph#23]
One task should be able to resolve multiple dependencies.

**Comment by k32** [comment]
After some reflection I think that this task is nothing but a different way to view the existing system, hence it can be implemented as a wrapper on top of the existing API. 

The current API basically describes the system in terms of interlocked actions, hence _task_ is the central concept and words _artefact_ and _source_ are not used even once. But conventional build systems are defined in terms of source files and artefacts and their task definition would look something like this:

```erlang
-type entity() :: source() | artefact().
 
#tg_task++{ requires := [entity(), ...]
          , provides := [entity(), ...]
          , action   := build_action()
          ...
          }
```

However, the above definition can be easily translated to the "old" style by trivially mapping entities to ids of tasks providing entities and a _provider_ function:

```erlang
%% Define rules how to build artefacts
-spec provider(entity()) -> #tg_task{} | undefined.
provider({source_code, Filename}) ->
  case filelib:is_file(Filename) of
    true -> undefined;
    false -> error({nofile, Filename})
  end;
provider({beam_file, Filename}) ->
  #tg_task{ id = {compile_beam, Filename ++ ".beam"}
          , data = #{compile_options =>[absolute_paths]}
          };
...
```

tl;dr: instead of trying to make artefacts and sources nodes of the graph, define "which task builds this entity" function and map it to `provides` and `requires` fields of #tg_task++.
